lib/charms/acme_client_operator/v0/acme_client.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# acme_client Library.
+
+This library is designed to enable developers to easily create new charms for the ACME protocol.
+This library contains all the logic necessary to get certificates from an ACME server..
+
+## Getting Started
+To get started using the library, you need to fetch the library using `charmcraft`.
+```shell
+charmcraft fetch-lib charms.acme_client_operator.v0.acme_client
+```
+You will also need to add the following library to the charm's `requirements.txt` file:
+- jsonschema
+- cryptography
+
+Then, to use the library in an example charm, you can do the following:
+```python
+from charms.acme_client_operator.v0.acme_client import AcmeClient
+from ops.main import main
+class ExampleAcmeCharm(AcmeClient):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._server = "https://acme-staging-v02.api.letsencrypt.org/directory"
+
+    @property
+    def _domain(self) -> Optional[str]:
+        return self.model.config.get("domain")
+
+    @property
+    def _email(self) -> Optional[str]:
+        return self.model.config.get("email")
+
+    @property
+    def _plugin(self) -> str:
+        return "namecheap"
+
+    @property
+    def _plugin_config(self):
+        return None
+```
+Charms that leverage this library also need to specify a `provides` relation in their
+`metadata.yaml` file. For example:
+```yaml
+provides:
+  certificates:
+    interface: tls-certificates
+```
+"""
+import abc
+import logging
+from abc import abstractmethod
+from typing import Dict, List
+
+from charms.tls_certificates_interface.v1.tls_certificates import (  # type: ignore[import]
+    CertificateCreationRequestEvent,
+    TLSCertificatesProvidesV1,
+)
+from cryptography import x509
+from cryptography.x509.oid import NameOID
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import ExecError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b3c9913b68dc42b89dfd0e77ac57236d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class AcmeClient(CharmBase):
```

Comment:
I would like to see a method in this class to handle the configuration of `self._server` and `self._email`. Ideally, it would look something like:

```
def update_generic_acme_config(self):
    # validate email and server in config, raising on error
    # set member variables
```

Then, the subclass could do something like this:

```
def _on_config_changed(self, event):
    ...
    try:
        self.update_generic_acme_config()
    except ValueError as e:
        self.unit.status = BlockedStatus(e)
        return
```

This would let us centralize the logic to handle those generics config and not have the validation code repeated in all subclasses. The charms would still need to specify the configuration in the metadata, but the code would be simplified.